src/NuGet.Core/NuGet.Packaging/Rules/IconMaxFilesizeRule.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.IO;
+using NuGet.Common;
+using System.CodeDom;
+using System.Globalization;
+using System.Runtime.CompilerServices;
+
+namespace NuGet.Packaging.Rules
+{
+    public class IconMaxFilesizeRule : IPackageRule
+    {
+        private const int MaxIconFilzeSize = 1024 * 1024;
+
+        private const int BufferSize = 1024 * 1024;
+
+        public string MessageFormat { get; }
+
+        public IconMaxFilesizeRule(string messageFormat)
+        {
+            MessageFormat = messageFormat ?? throw new ArgumentNullException(nameof(messageFormat));
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var reader = builder?.NuspecReader;
+
+            var path = reader?.GetIcon();
+
+            var issues = new List<PackagingLogMessage>();
+
+            if (!string.IsNullOrEmpty(path))
+            {
+                try
+                {
+                    using (var str = builder.GetStream(path))
+                    {
+                        long fileSize = EstimateFileSize(str);
+
+                        if (fileSize > MaxIconFilzeSize || (fileSize == MaxIconFilzeSize && str.CanRead))
+                        {
+                            issues.Add(PackagingLogMessage.CreateWarning(
+                                string.Format(CultureInfo.CurrentCulture, MessageFormat, path, AnalysisResources.IconMaxFilsesizeExceeded),
+                                NuGetLogCode.NU5037));
+                        }
+                    }
+                }
+                catch (FileNotFoundException e)
+                {
+                    issues.Add(PackagingLogMessage.CreateWarning(
+                                string.Format(CultureInfo.CurrentCulture, MessageFormat, path, e.Message),
+                                NuGetLogCode.NU5036));
+                }
+            }
+
+            return issues;
+        }
+
+        /// <summary>
+        /// Reads up to MaxIconFilzeSize of the file
+        /// </summary>
+        /// <param name="str"></param>
+        /// <returns></returns>
+        private static long EstimateFileSize(Stream str)
+        {
+            long fileSize = -1;
+            try
+            {
+                fileSize = str.Length;
+            }
+            catch(NotSupportedException)
```

Comment:
When I was doing the check in the Rules (after the package is created), I always hit the case in which you need to read the file content. Now, as a PackageBuilder methods, the file size is read from the stream.Length property. I'm considering keeping that codepath, just in case.

Reply:
I don't understand. Did you ever find an implementation of `Stream` that threw `NotSupportedException` on the property `Length` getter?

Reply:
Yes. I'm preparing an example.

Reply:
The validation now runs before creating the nupkg. This makes the file size available via the Stream class. Thus, there's no need for catching the exception